// @dsnp/sdk/dist/cjs/test/hardhatRPC.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
require("dotenv").config();
const ethers_1 = require("ethers");
const js_sha3_1 = require("js-sha3");
const publisher_1 = require("../core/contracts/publisher");
const config_1 = require("../config");
const hardhatRPC_1 = require("./hardhatRPC");
const TESTING_PRIVATE_KEY = String(process.env.TESTING_PRIVATE_KEY);
const RPC_URL = String(process.env.RPC_URL);
const provider = new ethers_1.ethers.providers.JsonRpcProvider(RPC_URL);
beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
    const oldConfig = yield config_1.getConfig();
    yield config_1.setConfig({
        contracts: {},
        provider: provider,
        signer: new ethers_1.ethers.Wallet(TESTING_PRIVATE_KEY, provider),
        store: oldConfig.store,
        queue: oldConfig.queue,
    });
}));
describe("snapshot and revert", () => {
    it("clears changes after revert", () => __awaiter(void 0, void 0, void 0, function* () {
        const filter = yield publisher_1.dsnpBatchFilter();
        jest.setTimeout(12000);
        yield hardhatRPC_1.snapshotHardhat(provider);
        const testUrl = "http://www.testconst.com";
        const hash = "0x" + js_sha3_1.keccak256("test");
        const publications = [{ dsnpType: 0, url: testUrl, hash: hash }];
        yield publisher_1.publish(publications);
        const batchEventLogs1 = yield provider.getLogs(filter);
        expect(batchEventLogs1.length).toEqual(1);
        yield hardhatRPC_1.revertHardhat(provider);
        const batchEventsLogs2 = yield provider.getLogs(filter);
        expect(batchEventsLogs2.length).toEqual(0);
    }));
});
