// @dsnp/sdk/dist/esm/config.js
export const MissingContract = new Error("Contract was not found");
export const MissingSigner = new Error("Signer is not set.");
export const MissingProvider = new Error("Blockchain provider is not set.");
export const MissingStore = new Error("Store adapter was not found");
export const MissingUser = new Error("No user id found. Please authenticate a handle.");
let config = {
    contracts: {},
};
export const getConfig = (overrides) => {
    if (!overrides)
        return config;
    return Object.assign(Object.assign(Object.assign({}, config), overrides), { contracts: Object.assign(Object.assign({}, config.contracts), overrides.contracts) });
};
export const setConfig = (newConfig) => {
    const { signer, provider } = newConfig;
    if (provider && signer && !signer.provider)
        newConfig.signer = signer.connect(provider);
    return (config = Object.assign(Object.assign({}, config), newConfig));
};
export const requireGetProvider = (opts) => {
    const c = getConfig(opts);
    if (!c.provider)
        throw MissingProvider;
    return c.provider;
};
export const requireGetSigner = (opts) => {
    const c = getConfig(opts);
    if (!c.signer)
        throw MissingSigner;
    return c.signer;
};
export const requireGetStore = (opts) => {
    const c = getConfig(opts);
    if (!c.store)
        throw MissingStore;
    return c.store;
};
export const requireGetCurrentFromId = (opts) => {
    const c = getConfig(opts);
    if (!c.currentFromId)
        throw MissingUser;
    return c.currentFromId;
};
export const getContracts = (opts) => {
    const c = getConfig(opts);
    return c.contracts;
};
