// @dsnp/sdk/dist/esm/content.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { keccak256 } from "js-sha3";
import * as activityPub from "./core/activityPub/activityPub";
import * as config from "./config";
import * as messages from "./core/messages/messages";
import { validateDSNPMessageId } from "./core/utilities";
import { requireGetStore } from "./config";
export const InvalidActivityPubOpts = new Error("Invalid activity pub options.");
export const InvalidInReplyTo = new Error("Invalid DSNP Message Id for inReplyTo");
export const broadcast = (contentOptions, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValid(contentObj))
        throw InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = keccak256(content);
    const store = requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createBroadcastMessage(currentFromId, url.toString(), contentHash);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
export const reply = (contentOptions, inReplyTo, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!validateDSNPMessageId(inReplyTo))
        throw InvalidInReplyTo;
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValidReply(contentObj))
        throw InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = keccak256(content);
    const store = requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createReplyMessage(currentFromId, url.toString(), contentHash, inReplyTo);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
export const react = (emoji, inReplyTo, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const currentFromId = config.requireGetCurrentFromId(opts);
    const message = messages.createReactionMessage(currentFromId, emoji, inReplyTo);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
export const profile = (contentOptions, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValidProfile(contentObj))
        throw InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = keccak256(content);
    const store = requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createProfileMessage(currentFromId, url.toString(), contentHash);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
