// @dsnp/sdk/dist/esm/index.js
import * as coreImport from "./core";
export const core = coreImport;
import * as generatorsImport from "./generators";
export const generators = generatorsImport;
export * from "./config";
export * from "./content";
export * from "./createPublication";
export * from "./handles";
export * from "./network";
export * from "./search";
import * as configImport from "./config";
export const config = configImport;
import * as contentImport from "./content";
export const content = contentImport;
import * as createPublicationImport from "./createPublication";
export const createPublication = createPublicationImport;
import * as handlesImport from "./handles";
export const handles = handlesImport;
import * as networkImport from "./network";
export const network = networkImport;
import * as searchImport from "./search";
export const search = searchImport;
export default Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, config), content), createPublication), handles), network), search), { core });
