// @dsnp/sdk/dist/cjs/generators/activityPubGenerators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateVideoAttachment = exports.generateImageAttachment = exports.generatePerson = exports.generateNote = exports.generateReply = void 0;
const activityPub_1 = require("../core/activityPub/activityPub");
const test_generators_1 = require("@dsnp/test-generators");
const PREFAB_FIRST_NAMES = test_generators_1.sampleText.prefabFirstNames;
const PREFAB_LAST_NAMES = test_generators_1.sampleText.prefabLastNames;
const PREFAB_URLS = test_generators_1.sampleText.prefabURLs;
const PREFAB_VIDEOS = test_generators_1.sampleText.prefabVideos;
const generateReply = (address, reply, inReplyTo) => {
    const note = exports.generateNote(address, reply, true);
    note.inReplyTo = inReplyTo;
    return note;
};
exports.generateReply = generateReply;
const generateNote = (address, message, hasAttachment) => {
    return activityPub_1.create({
        attributedTo: address,
        "@context": "https://www.w3.org/ns/activitystreams",
        content: message,
        attachments: hasAttachment ? [exports.generateImageAttachment()] : [],
        type: "Note",
    });
};
exports.generateNote = generateNote;
const generatePerson = (address, name) => {
    const newName = name ? name : [test_generators_1.sample(PREFAB_FIRST_NAMES), test_generators_1.sample(PREFAB_LAST_NAMES)].join(" ");
    return activityPub_1.create({
        attributedTo: address,
        "@context": "https://www.w3.org/ns/activitystreams",
        name: newName,
        url: "",
        type: "Person",
    });
};
exports.generatePerson = generatePerson;
const generateImageAttachment = (url) => {
    const theURL = url ? url : test_generators_1.sample(PREFAB_URLS);
    return {
        mediaType: theURL.replace(/(^\w+:|^)\/\//, ""),
        type: "Image",
        url: theURL,
    };
};
exports.generateImageAttachment = generateImageAttachment;
const generateVideoAttachment = (url) => {
    const theURL = url ? url : test_generators_1.sample(PREFAB_VIDEOS);
    return {
        mediaType: theURL.replace(/(^\w+:|^)\/\//, ""),
        type: "Video",
        url: theURL,
    };
};
exports.generateVideoAttachment = generateVideoAttachment;
