// @dsnp/sdk/dist/cjs/generators/batchFileGenerators.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateReactionBatchFile = exports.generateReplyBatchFile = exports.generateBroadcastBatchFile = exports.generateBatchReply = exports.generateBatchReaction = exports.generateBatchBroadcast = void 0;
const fs = __importStar(require("fs"));
const path_1 = __importDefault(require("path"));
const parquetjs_1 = __importDefault(require("@dsnp/parquetjs"));
const dsnpGenerators_1 = require("./dsnpGenerators");
const pq = __importStar(require("../core/batch/parquetSchema"));
const test_generators_1 = require("@dsnp/test-generators");
const generateBatchBroadcast = (from) => {
    return Object.assign({ signature: test_generators_1.generateHexString(256) }, dsnpGenerators_1.generateBroadcast(from));
};
exports.generateBatchBroadcast = generateBatchBroadcast;
const generateBatchReaction = (from) => {
    return Object.assign({ signature: test_generators_1.generateHexString(256) }, dsnpGenerators_1.generateReaction(from));
};
exports.generateBatchReaction = generateBatchReaction;
const generateBatchReply = (from) => {
    return Object.assign({ signature: test_generators_1.generateHexString(256) }, dsnpGenerators_1.generateReply(from));
};
exports.generateBatchReply = generateBatchReply;
const generateBroadcastBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.BroadcastSchema,
        generator: exports.generateBatchBroadcast,
        bloomOptions: pq.BroadcastBloomFilterOptions,
    });
});
exports.generateBroadcastBatchFile = generateBroadcastBatchFile;
const generateReplyBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.ReplySchema,
        generator: exports.generateBatchReply,
        bloomOptions: pq.ReplyBloomFilterOptions,
    });
});
exports.generateReplyBatchFile = generateReplyBatchFile;
const generateReactionBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.ReactionSchema,
        generator: exports.generateBatchReaction,
        bloomOptions: pq.ReactionBloomFilterOptions,
    });
});
exports.generateReactionBatchFile = generateReactionBatchFile;
const writeBatchFileWithOptions = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (opts.rootDir === "") {
        throw new Error("rootDir can't be blank");
    }
    ensureDir(opts.rootDir);
    const parquetFileName = ["replies", opts.numMessages, timestamp()].join("-") + ".parquet";
    const fname = path_1.default.join(opts.rootDir, parquetFileName);
    let itemsWritten = 0;
    try {
        const data = Array.from({ length: opts.numMessages }, () => opts.generator());
        const pSchema = new parquetjs_1.default.ParquetSchema(opts.schema);
        const writer = yield parquetjs_1.default.ParquetWriter.openFile(pSchema, fname, opts.bloomOptions);
        for (let i = 0; i < data.length; i++) {
            yield writer.appendRow(data[i]);
            itemsWritten++;
        }
        yield writer.close();
    }
    catch (e) {
        return { records: -1, path: fname, error: e.toString() };
    }
    return { records: itemsWritten, path: fname, error: "" };
});
const timestamp = () => {
    const now = new Date(Date.now());
    return [
        now.getUTCFullYear(),
        now.getUTCMonth(),
        now.getUTCDay(),
        "-",
        now.getUTCHours(),
        now.getUTCMinutes(),
        now.getUTCSeconds(),
        "-",
        now.getUTCMilliseconds(),
    ].join("");
};
const ensureDir = (dirname) => {
    if (!fs.existsSync(dirname)) {
        throw new Error(`directory does not exist: ${dirname}`);
    }
    else {
        const stats = fs.statSync(dirname);
        if (!stats.isDirectory()) {
            throw new Error(`is not a directory: ${dirname}`);
        }
    }
    return dirname;
};
