// @dsnp/sdk/dist/cjs/generators/dsnpGenerators.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateReaction = exports.generateReply = exports.generateBroadcast = exports.writeFixture = exports.generateDSNPStream = void 0;
const fs = __importStar(require("fs"));
const messages_1 = require("../core/messages/messages");
const test_generators_1 = require("@dsnp/test-generators");
const test_generators_2 = require("@dsnp/test-generators/");
const PREFAB_URLS = test_generators_2.sampleText.prefabURLs;
const generateEthereumAddress = test_generators_2.addresses.generateEthereumAddress;
const generateDSNPStream = (messageCount) => {
    const reactionReplyMax = 1000;
    const maxInt = reactionReplyMax * 10;
    return Array.from({ length: messageCount }, () => {
        const value = test_generators_1.randInt(maxInt);
        let msg;
        if (value > reactionReplyMax) {
            msg = value % 2 === 0 ? exports.generateReaction() : exports.generateReply();
        }
        else {
            msg = exports.generateBroadcast();
        }
        return msg;
    });
};
exports.generateDSNPStream = generateDSNPStream;
const writeFixture = (data, jsonFilePath) => {
    const ws = fs.createWriteStream(jsonFilePath).on("error", (e) => {
        throw new Error("createWriteStream failed: \n" + e.toString());
    });
    const ok = true;
    let i = 0;
    const header = '{ "entries": [';
    ws.write(header);
    do {
        ws.write(data[i] + ",");
        i++;
    } while (i < data.length - 1 && ok);
    ws.write(data[data.length - 1]);
    ws.end("]}");
    return ws.bytesWritten;
};
exports.writeFixture = writeFixture;
const generateBroadcast = (from) => {
    return {
        dsnpType: messages_1.DSNPType.Broadcast,
        fromId: from ? from : generateEthereumAddress(),
        contentHash: test_generators_1.generateHexString(64),
        url: test_generators_1.sample(PREFAB_URLS),
    };
};
exports.generateBroadcast = generateBroadcast;
const generateReply = (from) => {
    return {
        dsnpType: messages_1.DSNPType.Reply,
        fromId: from ? from : generateEthereumAddress(),
        inReplyTo: test_generators_1.generateHexString(64),
        contentHash: test_generators_1.generateHexString(64),
        url: test_generators_1.sample(PREFAB_URLS),
    };
};
exports.generateReply = generateReply;
const generateReaction = (from) => {
    return {
        dsnpType: messages_1.DSNPType.Reaction,
        fromId: from ? from : generateEthereumAddress(),
        emoji: test_generators_1.generateHexString(20),
        inReplyTo: test_generators_1.generateHexString(64),
    };
};
exports.generateReaction = generateReaction;
