// @dsnp/sdk/dist/esm/generators/activityPubGenerators.js
import { create } from "../core/activityPub/activityPub";
import { sample, sampleText } from "@dsnp/test-generators";
const PREFAB_FIRST_NAMES = sampleText.prefabFirstNames;
const PREFAB_LAST_NAMES = sampleText.prefabLastNames;
const PREFAB_URLS = sampleText.prefabURLs;
const PREFAB_VIDEOS = sampleText.prefabVideos;
export const generateReply = (address, reply, inReplyTo) => {
    const note = generateNote(address, reply, true);
    note.inReplyTo = inReplyTo;
    return note;
};
export const generateNote = (address, message, hasAttachment) => {
    return create({
        attributedTo: address,
        "@context": "https://www.w3.org/ns/activitystreams",
        content: message,
        attachments: hasAttachment ? [generateImageAttachment()] : [],
        type: "Note",
    });
};
export const generatePerson = (address, name) => {
    const newName = name ? name : [sample(PREFAB_FIRST_NAMES), sample(PREFAB_LAST_NAMES)].join(" ");
    return create({
        attributedTo: address,
        "@context": "https://www.w3.org/ns/activitystreams",
        name: newName,
        url: "",
        type: "Person",
    });
};
export const generateImageAttachment = (url) => {
    const theURL = url ? url : sample(PREFAB_URLS);
    return {
        mediaType: theURL.replace(/(^\w+:|^)\/\//, ""),
        type: "Image",
        url: theURL,
    };
};
export const generateVideoAttachment = (url) => {
    const theURL = url ? url : sample(PREFAB_VIDEOS);
    return {
        mediaType: theURL.replace(/(^\w+:|^)\/\//, ""),
        type: "Video",
        url: theURL,
    };
};
