// @dsnp/sdk/dist/esm/generators/batchFileGenerators.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fs from "fs";
import path from "path";
import parquet from "@dsnp/parquetjs";
import { generateBroadcast, generateReply, generateReaction } from "./dsnpGenerators";
import * as pq from "../core/batch/parquetSchema";
import { generateHexString } from "@dsnp/test-generators";
export const generateBatchBroadcast = (from) => {
    return Object.assign({ signature: generateHexString(256) }, generateBroadcast(from));
};
export const generateBatchReaction = (from) => {
    return Object.assign({ signature: generateHexString(256) }, generateReaction(from));
};
export const generateBatchReply = (from) => {
    return Object.assign({ signature: generateHexString(256) }, generateReply(from));
};
export const generateBroadcastBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.BroadcastSchema,
        generator: generateBatchBroadcast,
        bloomOptions: pq.BroadcastBloomFilterOptions,
    });
});
export const generateReplyBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.ReplySchema,
        generator: generateBatchReply,
        bloomOptions: pq.ReplyBloomFilterOptions,
    });
});
export const generateReactionBatchFile = (rootDir, numMessages) => __awaiter(void 0, void 0, void 0, function* () {
    return yield writeBatchFileWithOptions({
        rootDir: rootDir,
        numMessages: numMessages,
        schema: pq.ReactionSchema,
        generator: generateBatchReaction,
        bloomOptions: pq.ReactionBloomFilterOptions,
    });
});
const writeBatchFileWithOptions = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (opts.rootDir === "") {
        throw new Error("rootDir can't be blank");
    }
    ensureDir(opts.rootDir);
    const parquetFileName = ["replies", opts.numMessages, timestamp()].join("-") + ".parquet";
    const fname = path.join(opts.rootDir, parquetFileName);
    let itemsWritten = 0;
    try {
        const data = Array.from({ length: opts.numMessages }, () => opts.generator());
        const pSchema = new parquet.ParquetSchema(opts.schema);
        const writer = yield parquet.ParquetWriter.openFile(pSchema, fname, opts.bloomOptions);
        for (let i = 0; i < data.length; i++) {
            yield writer.appendRow(data[i]);
            itemsWritten++;
        }
        yield writer.close();
    }
    catch (e) {
        return { records: -1, path: fname, error: e.toString() };
    }
    return { records: itemsWritten, path: fname, error: "" };
});
const timestamp = () => {
    const now = new Date(Date.now());
    return [
        now.getUTCFullYear(),
        now.getUTCMonth(),
        now.getUTCDay(),
        "-",
        now.getUTCHours(),
        now.getUTCMinutes(),
        now.getUTCSeconds(),
        "-",
        now.getUTCMilliseconds(),
    ].join("");
};
const ensureDir = (dirname) => {
    if (!fs.existsSync(dirname)) {
        throw new Error(`directory does not exist: ${dirname}`);
    }
    else {
        const stats = fs.statSync(dirname);
        if (!stats.isDirectory()) {
            throw new Error(`is not a directory: ${dirname}`);
        }
    }
    return dirname;
};
