// @dsnp/sdk/dist/esm/generators/dsnpGenerators.js
import * as fs from "fs";
import { DSNPType } from "../core/messages/messages";
import { generateHexString, randInt, sample } from "@dsnp/test-generators";
import { addresses, sampleText } from "@dsnp/test-generators/";
const PREFAB_URLS = sampleText.prefabURLs;
const generateEthereumAddress = addresses.generateEthereumAddress;
export const generateDSNPStream = (messageCount) => {
    const reactionReplyMax = 1000;
    const maxInt = reactionReplyMax * 10;
    return Array.from({ length: messageCount }, () => {
        const value = randInt(maxInt);
        let msg;
        if (value > reactionReplyMax) {
            msg = value % 2 === 0 ? generateReaction() : generateReply();
        }
        else {
            msg = generateBroadcast();
        }
        return msg;
    });
};
export const writeFixture = (data, jsonFilePath) => {
    const ws = fs.createWriteStream(jsonFilePath).on("error", (e) => {
        throw new Error("createWriteStream failed: \n" + e.toString());
    });
    const ok = true;
    let i = 0;
    const header = '{ "entries": [';
    ws.write(header);
    do {
        ws.write(data[i] + ",");
        i++;
    } while (i < data.length - 1 && ok);
    ws.write(data[data.length - 1]);
    ws.end("]}");
    return ws.bytesWritten;
};
export const generateBroadcast = (from) => {
    return {
        dsnpType: DSNPType.Broadcast,
        fromId: from ? from : generateEthereumAddress(),
        contentHash: generateHexString(64),
        url: sample(PREFAB_URLS),
    };
};
export const generateReply = (from) => {
    return {
        dsnpType: DSNPType.Reply,
        fromId: from ? from : generateEthereumAddress(),
        inReplyTo: generateHexString(64),
        contentHash: generateHexString(64),
        url: sample(PREFAB_URLS),
    };
};
export const generateReaction = (from) => {
    return {
        dsnpType: DSNPType.Reaction,
        fromId: from ? from : generateEthereumAddress(),
        emoji: generateHexString(20),
        inReplyTo: generateHexString(64),
    };
};
