// @dsnp/sdk/dist/esm/core/batch/batch.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
import { ParquetReader, ParquetWriter, ParquetSchema } from "@dsnp/parquetjs";
import { keccak256 } from "js-sha3";
import { requireGetStore } from "../../config";
import { getSchemaFor, getBloomFilterOptionsFor } from "./parquetSchema";
import { MixedTypeBatchError, EmptyBatchError } from "../utilities";
export const createFile = (targetPath, messages, opts) => { var messages_1, messages_1_1; return __awaiter(void 0, void 0, void 0, function* () {
    var e_1, _a;
    let dsnpType;
    try {
        for (messages_1 = __asyncValues(messages); messages_1_1 = yield messages_1.next(), !messages_1_1.done;) {
            const message = messages_1_1.value;
            dsnpType = message.dsnpType;
            break;
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (messages_1_1 && !messages_1_1.done && (_a = messages_1.return)) yield _a.call(messages_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    if (dsnpType === undefined)
        throw new EmptyBatchError();
    const schema = new ParquetSchema(getSchemaFor(dsnpType));
    const bloomFilterOptions = getBloomFilterOptionsFor(dsnpType);
    const store = requireGetStore(opts);
    const hashGenerator = keccak256.create();
    const url = yield store.putStream(targetPath, (writeStream) => __awaiter(void 0, void 0, void 0, function* () {
        const hashingWriteStream = Object.assign(Object.assign({}, writeStream), { write: (chunk, ...args) => {
                hashGenerator.update(chunk);
                return writeStream.write(chunk, ...args);
            } });
        yield writeBatch(hashingWriteStream, schema, messages, bloomFilterOptions);
    }));
    return {
        url,
        hash: hashGenerator.hex(),
    };
}); };
export const writeBatch = (writeStream, schema, messages, opts) => { var messages_2, messages_2_1; return __awaiter(void 0, void 0, void 0, function* () {
    var e_2, _a;
    const writer = yield ParquetWriter.openStream(schema, writeStream, opts);
    let firstDsnpType;
    try {
        for (messages_2 = __asyncValues(messages); messages_2_1 = yield messages_2.next(), !messages_2_1.done;) {
            const message = messages_2_1.value;
            if (firstDsnpType === undefined)
                firstDsnpType = message.dsnpType;
            if (message.dsnpType != firstDsnpType)
                throw new MixedTypeBatchError(writeStream);
            writer.appendRow(message);
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (messages_2_1 && !messages_2_1.done && (_a = messages_2.return)) yield _a.call(messages_2);
        }
        finally { if (e_2) throw e_2.error; }
    }
    if (firstDsnpType === undefined)
        throw new EmptyBatchError(writeStream);
    yield writer.close();
}); };
export const openURL = (url) => __awaiter(void 0, void 0, void 0, function* () { return ParquetReader.openUrl(url); });
export const openFile = (path) => __awaiter(void 0, void 0, void 0, function* () { return ParquetReader.openFile(path); });
export const readFile = (reader, doReadRow) => __awaiter(void 0, void 0, void 0, function* () {
    const cursor = reader.getCursor();
    let record = null;
    while ((record = yield cursor.next())) {
        doReadRow(record);
    }
    return reader.close();
});
export const includes = (reader, column, item) => __awaiter(void 0, void 0, void 0, function* () {
    const bloomFilterData = yield reader.getBloomFilters([column]);
    return (bloomFilterData[column] || []).some((data) => data.sbbf.check(item));
});
