// @dsnp/sdk/dist/esm/core/batch/parquetSchema.js
import { DSNPType } from "../messages/messages";
export const BroadcastSchema = {
    dsnpType: { type: "INT32" },
    contentHash: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
export const BroadcastBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
export const ReplySchema = {
    dsnpType: { type: "INT32" },
    contentHash: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    inReplyTo: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
export const ReplyBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }, { column: "inReplyTo" }],
};
export const GraphChangeSchema = {
    dsnpType: { type: "INT32" },
    fromId: { type: "BYTE_ARRAY" },
    changeType: { type: "INT32" },
    signature: { type: "BYTE_ARRAY" },
};
export const GraphChangeBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
export const ProfileSchema = {
    dsnpType: { type: "INT32" },
    fromId: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
export const ProfileBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
export const ReactionSchema = {
    dsnpType: { type: "INT32" },
    emoji: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    inReplyTo: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
export const ReactionBloomFilterOptions = {
    bloomFilters: [{ column: "emoji" }, { column: "fromId" }, { column: "inReplyTo" }],
};
export const getSchemaFor = (dsnpType) => {
    switch (dsnpType) {
        case DSNPType.GraphChange:
            return GraphChangeSchema;
        case DSNPType.Broadcast:
            return BroadcastSchema;
        case DSNPType.Reply:
            return ReplySchema;
        case DSNPType.Reaction:
            return ReactionSchema;
        case DSNPType.Profile:
            return ProfileSchema;
    }
    throw new Error(`Invalid DSNP type: ${dsnpType}`);
};
export const getBloomFilterOptionsFor = (dsnpType) => {
    switch (dsnpType) {
        case DSNPType.GraphChange:
            return GraphChangeBloomFilterOptions;
        case DSNPType.Broadcast:
            return BroadcastBloomFilterOptions;
        case DSNPType.Reply:
            return ReplyBloomFilterOptions;
        case DSNPType.Reaction:
            return ReactionBloomFilterOptions;
        case DSNPType.Profile:
            return ProfileBloomFilterOptions;
    }
    throw new Error(`Invalid DSNP type: ${dsnpType}`);
};
