// @dsnp/sdk/dist/cjs/core/batch/batch.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const parquetjs_1 = require("@dsnp/parquetjs");
const batch = __importStar(require("./batch"));
const dsnpGenerators_1 = require("../../generators/dsnpGenerators");
const parquetSchema_1 = require("./parquetSchema");
const testStore_1 = __importDefault(require("../../test/testStore"));
const utilities_1 = require("../utilities");
describe("batch", () => {
    describe("includes", () => {
        let reader;
        beforeEach(() => {
            const MockSplitBlockBloomFilter = {
                check: jest.fn(),
            };
            const bloomFilterData = {
                name: [
                    {
                        sbbf: MockSplitBlockBloomFilter,
                        columnName: "name",
                        rowGroupIndex: 0,
                    },
                ],
            };
            reader = { getBloomFilters: jest.fn().mockResolvedValue(bloomFilterData) };
        });
        it("calls getBloomFilter", () => {
            batch.includes(reader, "name", "banana");
            expect(reader.getBloomFilters.mock.calls.length).toBe(1);
        });
    });
    describe("#writeBatch", () => {
        const parquetWriterInstance = {
            appendRow: jest.fn().mockImplementation((message) => __awaiter(void 0, void 0, void 0, function* () { return message; })),
            close: jest.fn(),
        };
        const writeStream = { write: jest.fn(), end: jest.fn() };
        const { writeBatch } = batch;
        beforeAll(() => {
            jest.spyOn(parquetjs_1.ParquetWriter, "openStream").mockResolvedValue(parquetWriterInstance);
        });
        beforeEach(jest.clearAllMocks);
        afterAll(jest.restoreAllMocks);
        describe("when passed a valid message iterator", () => {
            const messages = [Object.assign(Object.assign({}, dsnpGenerators_1.generateBroadcast()), { signature: "0xfa1ce" })];
            it("calls ParquetWriter#openStream to start a writable stream", () => __awaiter(void 0, void 0, void 0, function* () {
                yield writeBatch(writeStream, parquetSchema_1.BroadcastSchema, messages);
                expect(parquetjs_1.ParquetWriter.openStream).toHaveBeenCalled();
            }));
            it("calls ParquetWriter#appendRow to add a row to parquet stream", () => __awaiter(void 0, void 0, void 0, function* () {
                yield writeBatch(writeStream, parquetSchema_1.BroadcastSchema, messages);
                expect(parquetWriterInstance.appendRow).toHaveBeenCalledTimes(1);
            }));
            it("calls ParquetWriter#close to end the stream", () => __awaiter(void 0, void 0, void 0, function* () {
                yield writeBatch(writeStream, parquetSchema_1.BroadcastSchema, messages);
                expect(parquetWriterInstance.close).toHaveBeenCalledTimes(1);
            }));
        });
        describe("when passed a message iterator containing multiple DSNP types", () => {
            const badMessages = [
                Object.assign(Object.assign({}, dsnpGenerators_1.generateBroadcast()), { signature: "0xfa1ce" }),
                Object.assign(Object.assign({}, dsnpGenerators_1.generateReply()), { signature: "0xfa1ce" }),
            ];
            it("throws MixedDSNPTypeError", () => __awaiter(void 0, void 0, void 0, function* () {
                yield expect(writeBatch(writeStream, parquetSchema_1.BroadcastSchema, badMessages)).rejects.toBeInstanceOf(utilities_1.MixedTypeBatchError);
            }));
            it("includes the write stream handle in the thrown error", () => __awaiter(void 0, void 0, void 0, function* () {
                yield expect(writeBatch(writeStream, parquetSchema_1.BroadcastSchema, badMessages)).rejects.toMatchObject({
                    fileHandle: writeStream,
                });
            }));
        });
        describe("when passed a message iterator containing no messages", () => {
            const badMessages = [];
            it("throws MixedDSNPTypeError", () => __awaiter(void 0, void 0, void 0, function* () {
                yield expect(writeBatch(writeStream, parquetSchema_1.BroadcastSchema, badMessages)).rejects.toBeInstanceOf(utilities_1.EmptyBatchError);
            }));
            it("includes the write stream handle in the thrown error", () => __awaiter(void 0, void 0, void 0, function* () {
                yield expect(writeBatch(writeStream, parquetSchema_1.BroadcastSchema, badMessages)).rejects.toMatchObject({
                    fileHandle: writeStream,
                });
            }));
        });
    });
    describe("#createFile", () => {
        const { createFile } = batch;
        const messages = [Object.assign(Object.assign({}, dsnpGenerators_1.generateBroadcast()), { signature: "0xfa1ce" })];
        describe("when passed a valid message iterator", () => {
            it("calls putStream to start streaming", () => __awaiter(void 0, void 0, void 0, function* () {
                const mockStore = new testStore_1.default();
                jest.spyOn(mockStore, "putStream");
                yield createFile("batch.parquet", messages, { store: mockStore });
                expect(mockStore.putStream).toHaveBeenCalled();
            }));
            it("calls #writeBatch to stream write parquet", () => __awaiter(void 0, void 0, void 0, function* () {
                jest.spyOn(batch, "writeBatch");
                const mockStore = new testStore_1.default();
                yield createFile("batch.parquet", messages, { store: mockStore });
                const file = mockStore.getStore()["batch.parquet"];
                const reader = yield parquetjs_1.ParquetReader.openBuffer(file);
                expect(batch.writeBatch).toHaveBeenCalled();
                expect(reader.metadata.num_rows.buffer.toString("hex")).toEqual("0000000000000001");
            }));
        });
        describe("when passed a message iterator containing no messages", () => {
            const badMessages = [];
            it("throws MixedDSNPTypeError", () => __awaiter(void 0, void 0, void 0, function* () {
                const mockStore = new testStore_1.default();
                yield expect(createFile("batch.parquet", badMessages, { store: mockStore })).rejects.toBeInstanceOf(utilities_1.EmptyBatchError);
            }));
        });
    });
});
