// @dsnp/sdk/dist/cjs/core/batch/parquetSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBloomFilterOptionsFor = exports.getSchemaFor = exports.ReactionBloomFilterOptions = exports.ReactionSchema = exports.ProfileBloomFilterOptions = exports.ProfileSchema = exports.GraphChangeBloomFilterOptions = exports.GraphChangeSchema = exports.ReplyBloomFilterOptions = exports.ReplySchema = exports.BroadcastBloomFilterOptions = exports.BroadcastSchema = void 0;
const messages_1 = require("../messages/messages");
exports.BroadcastSchema = {
    dsnpType: { type: "INT32" },
    contentHash: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
exports.BroadcastBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
exports.ReplySchema = {
    dsnpType: { type: "INT32" },
    contentHash: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    inReplyTo: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
exports.ReplyBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }, { column: "inReplyTo" }],
};
exports.GraphChangeSchema = {
    dsnpType: { type: "INT32" },
    fromId: { type: "BYTE_ARRAY" },
    changeType: { type: "INT32" },
    signature: { type: "BYTE_ARRAY" },
};
exports.GraphChangeBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
exports.ProfileSchema = {
    dsnpType: { type: "INT32" },
    fromId: { type: "BYTE_ARRAY" },
    url: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
exports.ProfileBloomFilterOptions = {
    bloomFilters: [{ column: "fromId" }],
};
exports.ReactionSchema = {
    dsnpType: { type: "INT32" },
    emoji: { type: "BYTE_ARRAY" },
    fromId: { type: "BYTE_ARRAY" },
    inReplyTo: { type: "BYTE_ARRAY" },
    signature: { type: "BYTE_ARRAY" },
};
exports.ReactionBloomFilterOptions = {
    bloomFilters: [{ column: "emoji" }, { column: "fromId" }, { column: "inReplyTo" }],
};
const getSchemaFor = (dsnpType) => {
    switch (dsnpType) {
        case messages_1.DSNPType.GraphChange:
            return exports.GraphChangeSchema;
        case messages_1.DSNPType.Broadcast:
            return exports.BroadcastSchema;
        case messages_1.DSNPType.Reply:
            return exports.ReplySchema;
        case messages_1.DSNPType.Reaction:
            return exports.ReactionSchema;
        case messages_1.DSNPType.Profile:
            return exports.ProfileSchema;
    }
    throw new Error(`Invalid DSNP type: ${dsnpType}`);
};
exports.getSchemaFor = getSchemaFor;
const getBloomFilterOptionsFor = (dsnpType) => {
    switch (dsnpType) {
        case messages_1.DSNPType.GraphChange:
            return exports.GraphChangeBloomFilterOptions;
        case messages_1.DSNPType.Broadcast:
            return exports.BroadcastBloomFilterOptions;
        case messages_1.DSNPType.Reply:
            return exports.ReplyBloomFilterOptions;
        case messages_1.DSNPType.Reaction:
            return exports.ReactionBloomFilterOptions;
        case messages_1.DSNPType.Profile:
            return exports.ProfileBloomFilterOptions;
    }
    throw new Error(`Invalid DSNP type: ${dsnpType}`);
};
exports.getBloomFilterOptionsFor = getBloomFilterOptionsFor;
