// @dsnp/sdk/dist/esm/core/contracts/contract.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { keccak256 } from "js-sha3";
import * as types from "../../types/typechain";
import { ethers } from "ethers";
const DSNP_MIGRATION_TYPE = "DSNPMigration(address,string)";
export const getKeccakTopic = (topic) => "0x" + keccak256(topic);
const EVENTS_ABI = new ethers.utils.Interface([
    types.Publisher__factory,
    types.BeaconFactory__factory,
    types.Identity__factory,
    types.Migrations__factory,
    types.Registry__factory,
]
    .reduce((m, f) => m.concat(f.abi), [])
    .filter((ef) => ef.type === "event"));
export const DSNP_MIGRATION_ABI = [
    ethers.utils.ParamType.fromObject({
        indexed: false,
        baseType: "address",
        name: "contractAddr",
        type: "address",
    }),
    ethers.utils.ParamType.fromObject({
        indexed: false,
        baseType: "string",
        name: "contractName",
        type: "string",
    }),
];
const decodeReturnValues = (inputs, logs) => {
    const decoder = new ethers.utils.AbiCoder();
    return logs.map((log) => {
        const { contractAddr, contractName } = decoder.decode(inputs, log.data);
        return {
            contractAddr: contractAddr,
            contractName: contractName,
            blockNumber: log.blockNumber,
            blockHash: log.blockHash,
        };
    });
};
const filterValues = (values, contractName) => {
    return values.filter((result) => {
        return result.contractName == contractName;
    });
};
export const getContractAddress = (provider, contractName) => __awaiter(void 0, void 0, void 0, function* () {
    const topic = getKeccakTopic(DSNP_MIGRATION_TYPE);
    const logs = yield provider.getLogs({
        topics: [topic],
        fromBlock: 0,
    });
    const decodedValues = decodeReturnValues(DSNP_MIGRATION_ABI, logs);
    const filteredResults = filterValues(decodedValues, contractName);
    return filteredResults.length > 0 ? filteredResults[filteredResults.length - 1].contractAddr : null;
});
export const getVmError = (e) => {
    var _a, _b, _c;
    try {
        if (e.body) {
            const parsed = JSON.parse(e.body);
            return (_a = parsed === null || parsed === void 0 ? void 0 : parsed.error) === null || _a === void 0 ? void 0 : _a.message;
        }
        if ((_b = e.error) === null || _b === void 0 ? void 0 : _b.body) {
            const parsed = JSON.parse(e.error.body);
            return (_c = parsed === null || parsed === void 0 ? void 0 : parsed.error) === null || _c === void 0 ? void 0 : _c.message;
        }
    }
    catch (e) {
        return undefined;
    }
    return undefined;
};
export const parseLogs = (logs) => {
    return logs.map((log) => EVENTS_ABI.parseLog(log));
};
export const findEvent = (name, logs) => {
    const event = parseLogs(logs).find((e) => e.name === name);
    if (event === undefined) {
        throw `no ${name} logs found`;
    }
    return event;
};
