// @dsnp/sdk/dist/esm/core/contracts/identity.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { requireGetProvider, MissingProvider, MissingContract, requireGetSigner, getContracts, } from "../../config";
import { BeaconFactory__factory, Identity__factory, IdentityCloneFactory__factory, } from "../../types/typechain";
import { createTypedData, } from "./utilities";
import { getContractAddress } from "./contract";
const IDENTITY_CLONE_FACTORY_CONTRACT = "IdentityCloneFactory";
const IDENTITY_CONTRACT = "Identity";
const BEACON_FACTORY_CONTRACT = "BeaconFactory";
const BEACON_CONTRACT = "Beacon";
export var DelegationRole;
(function (DelegationRole) {
    DelegationRole[DelegationRole["NONE"] = 0] = "NONE";
    DelegationRole[DelegationRole["OWNER"] = 1] = "OWNER";
    DelegationRole[DelegationRole["ANNOUNCER"] = 2] = "ANNOUNCER";
})(DelegationRole || (DelegationRole = {}));
export var Permission;
(function (Permission) {
    Permission[Permission["NONE"] = 0] = "NONE";
    Permission[Permission["ANNOUNCE"] = 1] = "ANNOUNCE";
    Permission[Permission["OWNERSHIP_TRANSFER"] = 2] = "OWNERSHIP_TRANSFER";
    Permission[Permission["DELEGATE_ADD"] = 3] = "DELEGATE_ADD";
    Permission[Permission["DELEGATE_REMOVE"] = 4] = "DELEGATE_REMOVE";
})(Permission || (Permission = {}));
export const createCloneProxy = (logic, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!logic)
        logic = yield getIdentityLogicContractAddress(opts);
    const contract = yield getIdentityCloneFactoryContract(opts);
    return contract.createCloneProxy(logic);
});
export const createCloneProxyWithOwner = (owner, logic, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!logic)
        logic = yield getIdentityLogicContractAddress(opts);
    const contract = yield getIdentityCloneFactoryContract(opts);
    return contract.createCloneProxyWithOwner(logic, owner);
});
export const createBeaconProxy = (beacon, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getBeaconFactoryContract(opts);
    return contract["createBeaconProxy(address)"](beacon);
});
export const createBeaconProxyWithOwner = (owner, beacon, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getBeaconFactoryContract(opts);
    return contract.createBeaconProxyWithOwner(beacon, owner);
});
export const createAndRegisterBeaconProxy = (userAddress, handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const beaconFactory = yield getBeaconFactoryContract(opts);
    const beaconAddr = yield getBeaconAddress(opts);
    return yield beaconFactory.createAndRegisterBeaconProxy(beaconAddr, userAddress, handle);
});
const getIdentityLogicContractAddress = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { identityLogic } = getContracts(opts);
    const provider = requireGetProvider(opts);
    const address = identityLogic || (yield getContractAddress(provider, IDENTITY_CONTRACT));
    if (!address)
        throw MissingContract;
    return address;
});
const getIdentityCloneFactoryContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { identityCloneFactory } = getContracts(opts);
    const signer = requireGetSigner(opts);
    const provider = requireGetProvider(opts);
    const address = identityCloneFactory || (yield getContractAddress(provider, IDENTITY_CLONE_FACTORY_CONTRACT));
    if (!address)
        throw MissingContract;
    return IdentityCloneFactory__factory.connect(address, signer);
});
const getBeaconFactoryContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { beaconFactory } = getContracts(opts);
    const signer = requireGetSigner(opts);
    const provider = requireGetProvider(opts);
    const address = beaconFactory || (yield getContractAddress(provider, BEACON_FACTORY_CONTRACT));
    if (!address)
        throw MissingContract;
    return BeaconFactory__factory.connect(address, signer);
});
export const isAuthorizedTo = (address, contractAddress, permission, blockNumber, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = requireGetProvider(opts);
    if (!provider)
        throw MissingProvider;
    return Identity__factory.connect(contractAddress, provider).isAuthorizedTo(address, permission, blockNumber);
});
const getBeaconAddress = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { beacon } = getContracts(opts);
    const provider = requireGetProvider(opts);
    const address = beacon || (yield getContractAddress(provider, BEACON_CONTRACT));
    if (!address)
        throw MissingContract;
    return address;
});
export const getDomainSeparator = (contractAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = requireGetProvider(opts);
    return {
        name: "Identity",
        version: "1",
        chainId: (yield provider.getNetwork()).chainId,
        verifyingContract: contractAddress,
        salt: "0xa0bec69846cdcc8c1ba1eb93be1c5728385a9e26062a73e238b1beda189ac4c9",
    };
});
export const createAddDelegateEip712TypedData = (contractAddress, message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const primaryType = "DelegateAdd";
    const addDelegateType = [
        { name: "nonce", type: "uint32" },
        { name: "delegateAddr", type: "address" },
        { name: "role", type: "uint8" },
    ];
    const delegateAddMessage = Object.assign({ nonce: message.nonce || (yield getNonceForDelegate(contractAddress, message.delegateAddr)) }, message);
    const domainData = yield getDomainSeparator(contractAddress, opts);
    return createTypedData(domainData, primaryType, delegateAddMessage, { DelegateAdd: addDelegateType });
});
export const getNonceForDelegate = (contractAddress, delegateAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = requireGetProvider(opts);
    return yield Identity__factory.connect(contractAddress, provider).getNonceForDelegate(delegateAddress);
});
export const upsertDelegateBySignature = (contractAddress, signature, message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = requireGetSigner(opts);
    const provider = requireGetProvider(opts);
    const { v, r, s } = signature;
    const contract = yield Identity__factory.connect(contractAddress, provider);
    return contract.connect(signer).delegateByEIP712Sig(v, r, s, message);
});
export const upsertDelegate = (contractAddress, address, role, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = requireGetSigner(opts);
    const provider = requireGetProvider(opts);
    const contract = yield Identity__factory.connect(contractAddress, provider);
    return contract.connect(signer).delegate(address, role);
});
