// @dsnp/sdk/dist/esm/core/contracts/registry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ethers } from "ethers";
import { getContractAddress, getVmError } from "./contract";
import { MissingContract, requireGetSigner, requireGetProvider, getContracts } from "../../config";
import { Registry__factory } from "../../types/typechain";
import { resolveId } from "../../handles";
import { isAuthorizedTo } from "./identity";
import { serialize } from "../messages";
import { convertBigNumberToDSNPUserId, convertDSNPUserIdToBigNumber } from "../utilities/identifiers";
const CONTRACT_NAME = "Registry";
export const resolveRegistration = (handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    try {
        const [dsnpUserId, contractAddr] = yield contract.resolveRegistration(handle);
        return {
            handle,
            dsnpUserId: convertBigNumberToDSNPUserId(dsnpUserId),
            contractAddr,
        };
    }
    catch (e) {
        const error = e;
        const vmError = getVmError(error);
        if (vmError === null || vmError === void 0 ? void 0 : vmError.includes("Handle does not exist")) {
            return null;
        }
        throw e;
    }
});
export const register = (identityContractAddress, handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = requireGetSigner(opts);
    return yield contract.connect(signer).register(identityContractAddress, handle);
});
export const changeAddress = (handle, identityContractAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = requireGetSigner(opts);
    return yield contract.connect(signer).changeAddress(identityContractAddress, handle);
});
export const changeHandle = (oldHandle, newHandle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = requireGetSigner(opts);
    return yield contract.connect(signer).changeHandle(oldHandle, newHandle);
});
export const getDSNPRegistryUpdateEvents = (filter, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const dsnpUserId = filter.dsnpUserId ? convertDSNPUserIdToBigNumber(filter.dsnpUserId) : undefined;
    const contract = yield getContract(opts);
    const logs = yield contract.queryFilter(contract.filters.DSNPRegistryUpdate(dsnpUserId, filter.contractAddr));
    return logs.map((desc) => {
        const [id, addr, handle] = desc.args;
        const dsnpUserId = convertBigNumberToDSNPUserId(id);
        return { contractAddr: addr, dsnpUserId, handle };
    });
});
export const isMessageSignatureAuthorizedTo = (signature, message, dsnpUserId, permission, blockTag, opts) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const reg = yield resolveId(dsnpUserId);
    if (!reg)
        throw MissingContract;
    const provider = requireGetProvider(opts);
    let blockNumber = 0x0;
    if (blockTag) {
        const bn = (_a = (yield (provider === null || provider === void 0 ? void 0 : provider.getBlock(blockNumber)))) === null || _a === void 0 ? void 0 : _a.number;
        if (bn)
            blockNumber = bn;
    }
    const messageString = typeof message === "string" ? message : serialize(message);
    const signerAddr = ethers.utils.verifyMessage(messageString, signature);
    return isAuthorizedTo(signerAddr, reg.contractAddr, permission, blockNumber);
});
const getContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { registry } = getContracts(opts);
    const provider = requireGetProvider(opts);
    const address = registry || (yield getContractAddress(provider, CONTRACT_NAME));
    if (!address)
        throw MissingContract;
    return Registry__factory.connect(address, provider);
});
