// @dsnp/sdk/dist/esm/test/testStore.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
export default class TestStore {
    constructor() {
        this.store = {};
    }
    put(targetPath, content) {
        return __awaiter(this, void 0, void 0, function* () {
            this.store[targetPath] = content;
            return new URL(`http://fakestore.org/${targetPath}`);
        });
    }
    putStream(targetPath, callback) {
        return __awaiter(this, void 0, void 0, function* () {
            const buffers = [];
            const writeStream = {
                write: (...args) => {
                    const [chunk, maybeCallback1, maybeCallback2] = args;
                    if (typeof maybeCallback1 === "function")
                        maybeCallback1(null);
                    else if (typeof maybeCallback2 === "function")
                        maybeCallback2(null);
                    buffers.push(chunk);
                    return false;
                },
                end: (...args) => {
                    const [chunkOrFunction, maybeCallback] = args;
                    if (typeof chunkOrFunction === "function") {
                        chunkOrFunction(null);
                    }
                    else {
                        buffers.push(chunkOrFunction);
                        if (typeof maybeCallback === "function")
                            maybeCallback(null);
                    }
                },
            };
            yield callback(writeStream);
            this.store[targetPath] = Buffer.concat(buffers);
            return new URL(`http://fakestore.org/${targetPath}`);
        });
    }
    getStore() {
        return this.store;
    }
}
