// @dsnp/sdk/dist/cjs/core/utilities/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MixedTypeBatchError = exports.EmptyBatchError = exports.BatchError = exports.NotImplementedError = void 0;
exports.NotImplementedError = new Error("This method is not yet implemented.");
class BatchError extends Error {
    constructor(message, fileHandle) {
        super(message);
        this.name = "BatchError";
        this.fileHandle = fileHandle;
    }
}
exports.BatchError = BatchError;
class EmptyBatchError extends BatchError {
    constructor(fileHandle) {
        super("Invalid message iterator for batch: iterator contains no messages");
        this.name = "EmptyBatchError";
        this.fileHandle = fileHandle;
    }
}
exports.EmptyBatchError = EmptyBatchError;
class MixedTypeBatchError extends BatchError {
    constructor(fileHandle) {
        super("Invalid message iterator for batch: iterator contains multiple DSNP types", fileHandle);
        this.name = "MixedTypeBatchError";
    }
}
exports.MixedTypeBatchError = MixedTypeBatchError;
