// @dsnp/sdk/dist/cjs/types/typechain/factories/IDelegation__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IDelegation__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "delegate",
                type: "address",
            },
            {
                indexed: false,
                internalType: "enum IDelegation.Role",
                name: "role",
                type: "uint8",
            },
        ],
        name: "DSNPAddDelegate",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "delegate",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint64",
                name: "endBlock",
                type: "uint64",
            },
        ],
        name: "DSNPRemoveDelegate",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newDelegate",
                type: "address",
            },
            {
                internalType: "enum IDelegation.Role",
                name: "role",
                type: "uint8",
            },
        ],
        name: "delegate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "v",
                type: "uint8",
            },
            {
                internalType: "bytes32",
                name: "r",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "s",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "uint32",
                        name: "nonce",
                        type: "uint32",
                    },
                    {
                        internalType: "address",
                        name: "delegateAddr",
                        type: "address",
                    },
                    {
                        internalType: "enum IDelegation.Role",
                        name: "role",
                        type: "uint8",
                    },
                ],
                internalType: "struct IDelegation.DelegateAdd",
                name: "change",
                type: "tuple",
            },
        ],
        name: "delegateByEIP712Sig",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "addr",
                type: "address",
            },
            {
                internalType: "uint64",
                name: "endBlock",
                type: "uint64",
            },
        ],
        name: "delegateRemove",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "v",
                type: "uint8",
            },
            {
                internalType: "bytes32",
                name: "r",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "s",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "uint32",
                        name: "nonce",
                        type: "uint32",
                    },
                    {
                        internalType: "address",
                        name: "delegateAddr",
                        type: "address",
                    },
                    {
                        internalType: "uint64",
                        name: "endBlock",
                        type: "uint64",
                    },
                ],
                internalType: "struct IDelegation.DelegateRemove",
                name: "change",
                type: "tuple",
            },
        ],
        name: "delegateRemoveByEIP712Sig",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "addr",
                type: "address",
            },
        ],
        name: "getNonceForDelegate",
        outputs: [
            {
                internalType: "uint32",
                name: "",
                type: "uint32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "addr",
                type: "address",
            },
            {
                internalType: "enum IDelegation.Permission",
                name: "permission",
                type: "uint8",
            },
            {
                internalType: "uint256",
                name: "blockNumber",
                type: "uint256",
            },
        ],
        name: "isAuthorizedTo",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IDelegation__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IDelegation__factory = IDelegation__factory;
IDelegation__factory.abi = _abi;
