// @dsnp/sdk/dist/cjs/types/typechain/factories/IIdentityCloneFactory__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IIdentityCloneFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "addr",
                type: "address",
            },
        ],
        name: "ProxyCreated",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "logic",
                type: "address",
            },
        ],
        name: "createCloneProxy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "logic",
                type: "address",
            },
            {
                internalType: "address",
                name: "owner",
                type: "address",
            },
        ],
        name: "createCloneProxyWithOwner",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IIdentityCloneFactory__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IIdentityCloneFactory__factory = IIdentityCloneFactory__factory;
IIdentityCloneFactory__factory.abi = _abi;
