// @dsnp/sdk/dist/cjs/types/typechain/factories/IRegistry__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRegistry__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint64",
                name: "id",
                type: "uint64",
            },
            {
                indexed: true,
                internalType: "address",
                name: "addr",
                type: "address",
            },
            {
                indexed: false,
                internalType: "string",
                name: "handle",
                type: "string",
            },
        ],
        name: "DSNPRegistryUpdate",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newAddr",
                type: "address",
            },
            {
                internalType: "string",
                name: "handle",
                type: "string",
            },
        ],
        name: "changeAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "v",
                type: "uint8",
            },
            {
                internalType: "bytes32",
                name: "r",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "s",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "uint32",
                        name: "nonce",
                        type: "uint32",
                    },
                    {
                        internalType: "address",
                        name: "addr",
                        type: "address",
                    },
                    {
                        internalType: "string",
                        name: "handle",
                        type: "string",
                    },
                ],
                internalType: "struct IRegistry.AddressChange",
                name: "change",
                type: "tuple",
            },
        ],
        name: "changeAddressByEIP712Sig",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "oldHandle",
                type: "string",
            },
            {
                internalType: "string",
                name: "newHandle",
                type: "string",
            },
        ],
        name: "changeHandle",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "v",
                type: "uint8",
            },
            {
                internalType: "bytes32",
                name: "r",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "s",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "uint32",
                        name: "nonce",
                        type: "uint32",
                    },
                    {
                        internalType: "string",
                        name: "oldHandle",
                        type: "string",
                    },
                    {
                        internalType: "string",
                        name: "newHandle",
                        type: "string",
                    },
                ],
                internalType: "struct IRegistry.HandleChange",
                name: "change",
                type: "tuple",
            },
        ],
        name: "changeHandleByEIP712Sig",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "addr",
                type: "address",
            },
            {
                internalType: "string",
                name: "handle",
                type: "string",
            },
        ],
        name: "register",
        outputs: [
            {
                internalType: "uint64",
                name: "",
                type: "uint64",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "handle",
                type: "string",
            },
        ],
        name: "resolveHandleToNonce",
        outputs: [
            {
                internalType: "uint32",
                name: "",
                type: "uint32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "handle",
                type: "string",
            },
        ],
        name: "resolveRegistration",
        outputs: [
            {
                internalType: "uint64",
                name: "",
                type: "uint64",
            },
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IRegistry__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IRegistry__factory = IRegistry__factory;
IRegistry__factory.abi = _abi;
