// @dsnp/sdk/dist/cjs/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContracts = exports.requireGetCurrentFromId = exports.requireGetStore = exports.requireGetSigner = exports.requireGetProvider = exports.setConfig = exports.getConfig = exports.MissingUser = exports.MissingStore = exports.MissingProvider = exports.MissingSigner = exports.MissingContract = void 0;
exports.MissingContract = new Error("Contract was not found");
exports.MissingSigner = new Error("Signer is not set.");
exports.MissingProvider = new Error("Blockchain provider is not set.");
exports.MissingStore = new Error("Store adapter was not found");
exports.MissingUser = new Error("No user id found. Please authenticate a handle.");
let config = {
    contracts: {},
};
const getConfig = (overrides) => {
    if (!overrides)
        return config;
    return Object.assign(Object.assign(Object.assign({}, config), overrides), { contracts: Object.assign(Object.assign({}, config.contracts), overrides.contracts) });
};
exports.getConfig = getConfig;
const setConfig = (newConfig) => {
    const { signer, provider } = newConfig;
    if (provider && signer && !signer.provider)
        newConfig.signer = signer.connect(provider);
    return (config = Object.assign(Object.assign({}, config), newConfig));
};
exports.setConfig = setConfig;
const requireGetProvider = (opts) => {
    const c = exports.getConfig(opts);
    if (!c.provider)
        throw exports.MissingProvider;
    return c.provider;
};
exports.requireGetProvider = requireGetProvider;
const requireGetSigner = (opts) => {
    const c = exports.getConfig(opts);
    if (!c.signer)
        throw exports.MissingSigner;
    return c.signer;
};
exports.requireGetSigner = requireGetSigner;
const requireGetStore = (opts) => {
    const c = exports.getConfig(opts);
    if (!c.store)
        throw exports.MissingStore;
    return c.store;
};
exports.requireGetStore = requireGetStore;
const requireGetCurrentFromId = (opts) => {
    const c = exports.getConfig(opts);
    if (!c.currentFromId)
        throw exports.MissingUser;
    return c.currentFromId;
};
exports.requireGetCurrentFromId = requireGetCurrentFromId;
const getContracts = (opts) => {
    const c = exports.getConfig(opts);
    return c.contracts;
};
exports.getContracts = getContracts;
