// @dsnp/sdk/dist/cjs/content.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.profile = exports.react = exports.reply = exports.broadcast = exports.InvalidInReplyTo = exports.InvalidActivityPubOpts = void 0;
const js_sha3_1 = require("js-sha3");
const activityPub = __importStar(require("./core/activityPub/activityPub"));
const config = __importStar(require("./config"));
const messages = __importStar(require("./core/messages/messages"));
const utilities_1 = require("./core/utilities");
const config_1 = require("./config");
exports.InvalidActivityPubOpts = new Error("Invalid activity pub options.");
exports.InvalidInReplyTo = new Error("Invalid DSNP Message Id for inReplyTo");
const broadcast = (contentOptions, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValid(contentObj))
        throw exports.InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = js_sha3_1.keccak256(content);
    const store = config_1.requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createBroadcastMessage(currentFromId, url.toString(), contentHash);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
exports.broadcast = broadcast;
const reply = (contentOptions, inReplyTo, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!utilities_1.validateDSNPMessageId(inReplyTo))
        throw exports.InvalidInReplyTo;
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValidReply(contentObj))
        throw exports.InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = js_sha3_1.keccak256(content);
    const store = config_1.requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createReplyMessage(currentFromId, url.toString(), contentHash, inReplyTo);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
exports.reply = reply;
const react = (emoji, inReplyTo, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const currentFromId = config.requireGetCurrentFromId(opts);
    const message = messages.createReactionMessage(currentFromId, emoji, inReplyTo);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
exports.react = react;
const profile = (contentOptions, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contentObj = activityPub.create(contentOptions);
    if (!activityPub.isValidProfile(contentObj))
        throw exports.InvalidActivityPubOpts;
    const content = activityPub.serialize(contentObj);
    const currentFromId = config.requireGetCurrentFromId(opts);
    const contentHash = js_sha3_1.keccak256(content);
    const store = config_1.requireGetStore(opts);
    const url = yield store.put(contentHash, content);
    const message = messages.createProfileMessage(currentFromId, url.toString(), contentHash);
    const signedMessage = yield messages.sign(message, opts);
    return signedMessage;
});
exports.profile = profile;
