// @dsnp/sdk/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.search = exports.network = exports.handles = exports.createPublication = exports.content = exports.config = exports.generators = exports.core = void 0;
const coreImport = __importStar(require("./core"));
exports.core = coreImport;
const generatorsImport = __importStar(require("./generators"));
exports.generators = generatorsImport;
__exportStar(require("./config"), exports);
__exportStar(require("./content"), exports);
__exportStar(require("./createPublication"), exports);
__exportStar(require("./handles"), exports);
__exportStar(require("./network"), exports);
__exportStar(require("./search"), exports);
const configImport = __importStar(require("./config"));
exports.config = configImport;
const contentImport = __importStar(require("./content"));
exports.content = contentImport;
const createPublicationImport = __importStar(require("./createPublication"));
exports.createPublication = createPublicationImport;
const handlesImport = __importStar(require("./handles"));
exports.handles = handlesImport;
const networkImport = __importStar(require("./network"));
exports.network = networkImport;
const searchImport = __importStar(require("./search"));
exports.search = searchImport;
exports.default = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, exports.config), exports.content), exports.createPublication), exports.handles), exports.network), exports.search), { core: exports.core });
