// @dsnp/sdk/dist/cjs/network.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const config = __importStar(require("./config"));
const network = __importStar(require("./network"));
const contract_1 = require("./core/contracts/contract");
const registry_1 = require("./core/contracts/registry");
const messages_1 = require("./core/messages/messages");
const typechain_1 = require("./types/typechain");
const sdkTestConfig_1 = require("./test/sdkTestConfig");
const hardhatRPC_1 = require("./test/hardhatRPC");
const identifiers_1 = require("./core/utilities/identifiers");
describe("network", () => {
    let registerId;
    hardhatRPC_1.setupSnapshot();
    const { signer, provider } = sdkTestConfig_1.setupConfig();
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        yield hardhatRPC_1.snapshotHardhat(provider);
        const fakeAddress = "0x1Ea32de10D5a18e55DEBAf379B26Cc0c6952B168";
        const identityContract = yield new typechain_1.Identity__factory(signer).deploy(fakeAddress);
        yield identityContract.deployed();
        const txn = yield registry_1.register(identityContract.address, "dril");
        const receipt = yield txn.wait(1);
        const registerEvent = contract_1.findEvent("DSNPRegistryUpdate", receipt.logs);
        registerId = identifiers_1.convertBigNumberToDSNPUserId(registerEvent.args[0]);
    }));
    afterAll(() => __awaiter(void 0, void 0, void 0, function* () {
        yield hardhatRPC_1.revertHardhat(provider);
    }));
    describe("follow", () => {
        describe("with a valid signer and user id", () => {
            beforeEach(() => {
                config.setConfig({
                    currentFromId: "dsnp://0000000000000000",
                    signer,
                    provider,
                });
            });
            it("returns a follow graph change DSNP message", () => __awaiter(void 0, void 0, void 0, function* () {
                const message = yield network.follow(registerId);
                expect(message).toMatchObject({
                    fromId: "dsnp://0000000000000000",
                    dsnpType: messages_1.DSNPType.GraphChange,
                    changeType: messages_1.DSNPGraphChangeType.Follow,
                    objectId: registerId,
                });
            }));
        });
        describe("without a signer", () => {
            it("throws MissingUser", () => __awaiter(void 0, void 0, void 0, function* () {
                config.setConfig({
                    currentFromId: "dsnp://0000000000000000",
                    signer: undefined,
                    provider,
                });
                yield expect(network.follow(registerId)).rejects.toThrow(config.MissingSigner);
            }));
        });
        describe("without a user id", () => {
            it("throws MissingUser", () => __awaiter(void 0, void 0, void 0, function* () {
                config.setConfig({
                    currentFromId: undefined,
                    signer,
                    provider,
                });
                yield expect(network.follow("dril")).rejects.toThrow(config.MissingUser);
            }));
        });
    });
    describe("unfollow", () => {
        describe("with a valid signer and user id", () => {
            beforeEach(() => {
                config.setConfig({
                    currentFromId: "dsnp://0000000000000000",
                    signer,
                    provider,
                });
            });
            it("returns a follow graph change DSNP message", () => __awaiter(void 0, void 0, void 0, function* () {
                const message = yield network.unfollow(registerId);
                expect(message).toMatchObject({
                    fromId: "dsnp://0000000000000000",
                    dsnpType: messages_1.DSNPType.GraphChange,
                    changeType: messages_1.DSNPGraphChangeType.Follow,
                    objectId: registerId,
                });
            }));
        });
        describe("without a signer", () => {
            it("throws MissingUser", () => __awaiter(void 0, void 0, void 0, function* () {
                config.setConfig({
                    currentFromId: "dsnp://0000000000000000",
                    signer: undefined,
                    provider,
                });
                yield expect(network.unfollow(registerId)).rejects.toThrow(config.MissingSigner);
            }));
        });
        describe("without a user id", () => {
            it("throws MissingUser", () => __awaiter(void 0, void 0, void 0, function* () {
                config.setConfig({
                    currentFromId: undefined,
                    signer,
                    provider,
                });
                yield expect(network.unfollow(registerId)).rejects.toThrow(config.MissingUser);
            }));
        });
    });
});
