// @dsnp/sdk/dist/cjs/core/contracts/identity.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.upsertDelegate = exports.upsertDelegateBySignature = exports.getNonceForDelegate = exports.createAddDelegateEip712TypedData = exports.getDomainSeparator = exports.isAuthorizedTo = exports.createAndRegisterBeaconProxy = exports.createBeaconProxyWithOwner = exports.createBeaconProxy = exports.createCloneProxyWithOwner = exports.createCloneProxy = exports.Permission = exports.DelegationRole = void 0;
const config_1 = require("../../config");
const typechain_1 = require("../../types/typechain");
const utilities_1 = require("./utilities");
const contract_1 = require("./contract");
const IDENTITY_CLONE_FACTORY_CONTRACT = "IdentityCloneFactory";
const IDENTITY_CONTRACT = "Identity";
const BEACON_FACTORY_CONTRACT = "BeaconFactory";
const BEACON_CONTRACT = "Beacon";
var DelegationRole;
(function (DelegationRole) {
    DelegationRole[DelegationRole["NONE"] = 0] = "NONE";
    DelegationRole[DelegationRole["OWNER"] = 1] = "OWNER";
    DelegationRole[DelegationRole["ANNOUNCER"] = 2] = "ANNOUNCER";
})(DelegationRole = exports.DelegationRole || (exports.DelegationRole = {}));
var Permission;
(function (Permission) {
    Permission[Permission["NONE"] = 0] = "NONE";
    Permission[Permission["ANNOUNCE"] = 1] = "ANNOUNCE";
    Permission[Permission["OWNERSHIP_TRANSFER"] = 2] = "OWNERSHIP_TRANSFER";
    Permission[Permission["DELEGATE_ADD"] = 3] = "DELEGATE_ADD";
    Permission[Permission["DELEGATE_REMOVE"] = 4] = "DELEGATE_REMOVE";
})(Permission = exports.Permission || (exports.Permission = {}));
const createCloneProxy = (logic, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!logic)
        logic = yield getIdentityLogicContractAddress(opts);
    const contract = yield getIdentityCloneFactoryContract(opts);
    return contract.createCloneProxy(logic);
});
exports.createCloneProxy = createCloneProxy;
const createCloneProxyWithOwner = (owner, logic, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (!logic)
        logic = yield getIdentityLogicContractAddress(opts);
    const contract = yield getIdentityCloneFactoryContract(opts);
    return contract.createCloneProxyWithOwner(logic, owner);
});
exports.createCloneProxyWithOwner = createCloneProxyWithOwner;
const createBeaconProxy = (beacon, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getBeaconFactoryContract(opts);
    return contract["createBeaconProxy(address)"](beacon);
});
exports.createBeaconProxy = createBeaconProxy;
const createBeaconProxyWithOwner = (owner, beacon, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getBeaconFactoryContract(opts);
    return contract.createBeaconProxyWithOwner(beacon, owner);
});
exports.createBeaconProxyWithOwner = createBeaconProxyWithOwner;
const createAndRegisterBeaconProxy = (userAddress, handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const beaconFactory = yield getBeaconFactoryContract(opts);
    const beaconAddr = yield getBeaconAddress(opts);
    return yield beaconFactory.createAndRegisterBeaconProxy(beaconAddr, userAddress, handle);
});
exports.createAndRegisterBeaconProxy = createAndRegisterBeaconProxy;
const getIdentityLogicContractAddress = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { identityLogic } = config_1.getContracts(opts);
    const provider = config_1.requireGetProvider(opts);
    const address = identityLogic || (yield contract_1.getContractAddress(provider, IDENTITY_CONTRACT));
    if (!address)
        throw config_1.MissingContract;
    return address;
});
const getIdentityCloneFactoryContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { identityCloneFactory } = config_1.getContracts(opts);
    const signer = config_1.requireGetSigner(opts);
    const provider = config_1.requireGetProvider(opts);
    const address = identityCloneFactory || (yield contract_1.getContractAddress(provider, IDENTITY_CLONE_FACTORY_CONTRACT));
    if (!address)
        throw config_1.MissingContract;
    return typechain_1.IdentityCloneFactory__factory.connect(address, signer);
});
const getBeaconFactoryContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { beaconFactory } = config_1.getContracts(opts);
    const signer = config_1.requireGetSigner(opts);
    const provider = config_1.requireGetProvider(opts);
    const address = beaconFactory || (yield contract_1.getContractAddress(provider, BEACON_FACTORY_CONTRACT));
    if (!address)
        throw config_1.MissingContract;
    return typechain_1.BeaconFactory__factory.connect(address, signer);
});
const isAuthorizedTo = (address, contractAddress, permission, blockNumber, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = config_1.requireGetProvider(opts);
    if (!provider)
        throw config_1.MissingProvider;
    return typechain_1.Identity__factory.connect(contractAddress, provider).isAuthorizedTo(address, permission, blockNumber);
});
exports.isAuthorizedTo = isAuthorizedTo;
const getBeaconAddress = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { beacon } = config_1.getContracts(opts);
    const provider = config_1.requireGetProvider(opts);
    const address = beacon || (yield contract_1.getContractAddress(provider, BEACON_CONTRACT));
    if (!address)
        throw config_1.MissingContract;
    return address;
});
const getDomainSeparator = (contractAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = config_1.requireGetProvider(opts);
    return {
        name: "Identity",
        version: "1",
        chainId: (yield provider.getNetwork()).chainId,
        verifyingContract: contractAddress,
        salt: "0xa0bec69846cdcc8c1ba1eb93be1c5728385a9e26062a73e238b1beda189ac4c9",
    };
});
exports.getDomainSeparator = getDomainSeparator;
const createAddDelegateEip712TypedData = (contractAddress, message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const primaryType = "DelegateAdd";
    const addDelegateType = [
        { name: "nonce", type: "uint32" },
        { name: "delegateAddr", type: "address" },
        { name: "role", type: "uint8" },
    ];
    const delegateAddMessage = Object.assign({ nonce: message.nonce || (yield exports.getNonceForDelegate(contractAddress, message.delegateAddr)) }, message);
    const domainData = yield exports.getDomainSeparator(contractAddress, opts);
    return utilities_1.createTypedData(domainData, primaryType, delegateAddMessage, { DelegateAdd: addDelegateType });
});
exports.createAddDelegateEip712TypedData = createAddDelegateEip712TypedData;
const getNonceForDelegate = (contractAddress, delegateAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = config_1.requireGetProvider(opts);
    return yield typechain_1.Identity__factory.connect(contractAddress, provider).getNonceForDelegate(delegateAddress);
});
exports.getNonceForDelegate = getNonceForDelegate;
const upsertDelegateBySignature = (contractAddress, signature, message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = config_1.requireGetSigner(opts);
    const provider = config_1.requireGetProvider(opts);
    const { v, r, s } = signature;
    const contract = yield typechain_1.Identity__factory.connect(contractAddress, provider);
    return contract.connect(signer).delegateByEIP712Sig(v, r, s, message);
});
exports.upsertDelegateBySignature = upsertDelegateBySignature;
const upsertDelegate = (contractAddress, address, role, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = config_1.requireGetSigner(opts);
    const provider = config_1.requireGetProvider(opts);
    const contract = yield typechain_1.Identity__factory.connect(contractAddress, provider);
    return contract.connect(signer).delegate(address, role);
});
exports.upsertDelegate = upsertDelegate;
