// @dsnp/sdk/dist/cjs/core/contracts/registry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMessageSignatureAuthorizedTo = exports.getDSNPRegistryUpdateEvents = exports.changeHandle = exports.changeAddress = exports.register = exports.resolveRegistration = void 0;
const ethers_1 = require("ethers");
const contract_1 = require("./contract");
const config_1 = require("../../config");
const typechain_1 = require("../../types/typechain");
const handles_1 = require("../../handles");
const identity_1 = require("./identity");
const messages_1 = require("../messages");
const identifiers_1 = require("../utilities/identifiers");
const CONTRACT_NAME = "Registry";
const resolveRegistration = (handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    try {
        const [dsnpUserId, contractAddr] = yield contract.resolveRegistration(handle);
        return {
            handle,
            dsnpUserId: identifiers_1.convertBigNumberToDSNPUserId(dsnpUserId),
            contractAddr,
        };
    }
    catch (e) {
        const error = e;
        const vmError = contract_1.getVmError(error);
        if (vmError === null || vmError === void 0 ? void 0 : vmError.includes("Handle does not exist")) {
            return null;
        }
        throw e;
    }
});
exports.resolveRegistration = resolveRegistration;
const register = (identityContractAddress, handle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = config_1.requireGetSigner(opts);
    return yield contract.connect(signer).register(identityContractAddress, handle);
});
exports.register = register;
const changeAddress = (handle, identityContractAddress, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = config_1.requireGetSigner(opts);
    return yield contract.connect(signer).changeAddress(identityContractAddress, handle);
});
exports.changeAddress = changeAddress;
const changeHandle = (oldHandle, newHandle, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const contract = yield getContract(opts);
    const signer = config_1.requireGetSigner(opts);
    return yield contract.connect(signer).changeHandle(oldHandle, newHandle);
});
exports.changeHandle = changeHandle;
const getDSNPRegistryUpdateEvents = (filter, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const dsnpUserId = filter.dsnpUserId ? identifiers_1.convertDSNPUserIdToBigNumber(filter.dsnpUserId) : undefined;
    const contract = yield getContract(opts);
    const logs = yield contract.queryFilter(contract.filters.DSNPRegistryUpdate(dsnpUserId, filter.contractAddr));
    return logs.map((desc) => {
        const [id, addr, handle] = desc.args;
        const dsnpUserId = identifiers_1.convertBigNumberToDSNPUserId(id);
        return { contractAddr: addr, dsnpUserId, handle };
    });
});
exports.getDSNPRegistryUpdateEvents = getDSNPRegistryUpdateEvents;
const isMessageSignatureAuthorizedTo = (signature, message, dsnpUserId, permission, blockTag, opts) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const reg = yield handles_1.resolveId(dsnpUserId);
    if (!reg)
        throw config_1.MissingContract;
    const provider = config_1.requireGetProvider(opts);
    let blockNumber = 0x0;
    if (blockTag) {
        const bn = (_a = (yield (provider === null || provider === void 0 ? void 0 : provider.getBlock(blockNumber)))) === null || _a === void 0 ? void 0 : _a.number;
        if (bn)
            blockNumber = bn;
    }
    const messageString = typeof message === "string" ? message : messages_1.serialize(message);
    const signerAddr = ethers_1.ethers.utils.verifyMessage(messageString, signature);
    return identity_1.isAuthorizedTo(signerAddr, reg.contractAddr, permission, blockNumber);
});
exports.isMessageSignatureAuthorizedTo = isMessageSignatureAuthorizedTo;
const getContract = (opts) => __awaiter(void 0, void 0, void 0, function* () {
    const { registry } = config_1.getContracts(opts);
    const provider = config_1.requireGetProvider(opts);
    const address = registry || (yield contract_1.getContractAddress(provider, CONTRACT_NAME));
    if (!address)
        throw config_1.MissingContract;
    return typechain_1.Registry__factory.connect(address, provider);
});
