// @dsnp/sdk/dist/cjs/core/contracts/subscription.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscribeToBatchPublications = void 0;
const ethers_1 = require("ethers");
const config_1 = require("../../config");
const publisher_1 = require("./publisher");
const typechain_1 = require("../../types/typechain");
const PUBLISHER_DECODER = new ethers_1.ethers.utils.Interface(typechain_1.Publisher__factory.abi);
const subscribeToBatchPublications = (doReceivePublication, filter) => __awaiter(void 0, void 0, void 0, function* () {
    let pastLogs = [];
    const currentLogQueue = [];
    const batchFilter = yield publisher_1.dsnpBatchFilter();
    const batchFilterWithOptions = filter ? createFilter(batchFilter, filter) : batchFilter;
    const provider = config_1.requireGetProvider();
    let maxBlockNumberForPastLogs = (filter === null || filter === void 0 ? void 0 : filter.fromBlock) || 0;
    let useQueue = (filter === null || filter === void 0 ? void 0 : filter.fromBlock) != undefined;
    provider.on(batchFilterWithOptions, (log) => {
        const logItem = decodeLogsForBatchPublication([log])[0];
        if (useQueue) {
            currentLogQueue.push(logItem);
        }
        else if (logItem.blockNumber > maxBlockNumberForPastLogs) {
            doReceivePublication(logItem);
        }
    });
    if (useQueue) {
        pastLogs = yield getPastLogs(provider, { fromBlock: filter === null || filter === void 0 ? void 0 : filter.fromBlock });
        maxBlockNumberForPastLogs = pastLogs[pastLogs.length - 1].blockNumber;
        while (pastLogs.length > 0) {
            const batchItem = pastLogs.shift();
            if (batchItem)
                doReceivePublication(batchItem);
        }
        while (currentLogQueue.length > 0) {
            const batchItem = currentLogQueue.shift();
            if (batchItem && batchItem.blockNumber > maxBlockNumberForPastLogs)
                doReceivePublication(batchItem);
        }
        useQueue = false;
    }
    return () => {
        provider.off(batchFilterWithOptions);
    };
});
exports.subscribeToBatchPublications = subscribeToBatchPublications;
const createFilter = (batchFilter, filterOptions) => {
    const topics = batchFilter.topics ? batchFilter.topics : [];
    const dsnpTypeTopic = (filterOptions === null || filterOptions === void 0 ? void 0 : filterOptions.dsnpType) ? "0x" + filterOptions.dsnpType.toString(16).padStart(64, "0") : null;
    if (dsnpTypeTopic) {
        topics.push(dsnpTypeTopic);
    }
    const finalFilter = {
        topics: topics,
    };
    return finalFilter;
};
const getPastLogs = (provider, filter) => __awaiter(void 0, void 0, void 0, function* () {
    const logs = yield provider.getLogs(filter);
    return decodeLogsForBatchPublication(logs);
});
const decodeLogsForBatchPublication = (logs) => {
    return logs
        .map((log) => {
        const fragment = PUBLISHER_DECODER.parseLog(log);
        return { fragment, log: log };
    })
        .filter((desc) => desc.fragment.name === "DSNPBatchPublication")
        .map((item) => {
        return {
            dsnpType: item.fragment.args.dsnpType,
            dsnpHash: item.fragment.args.dsnpHash,
            dsnpUrl: item.fragment.args.dsnpUrl,
            blockNumber: item.log.blockNumber,
            transactionHash: item.log.transactionHash,
        };
    });
};
