// @dsnp/sdk/dist/esm/types/typechain/factories/IIdentityCloneFactory__factory.js
import { Contract, utils } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "addr",
                type: "address",
            },
        ],
        name: "ProxyCreated",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "logic",
                type: "address",
            },
        ],
        name: "createCloneProxy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "logic",
                type: "address",
            },
            {
                internalType: "address",
                name: "owner",
                type: "address",
            },
        ],
        name: "createCloneProxyWithOwner",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IIdentityCloneFactory__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
IIdentityCloneFactory__factory.abi = _abi;
