// @dsnp/sdk/dist/esm/types/typechain/factories/IPublish__factory.js
import { Contract, utils } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "int16",
                name: "dsnpType",
                type: "int16",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "dsnpHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "dsnpUrl",
                type: "string",
            },
        ],
        name: "DSNPBatchPublication",
        type: "event",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "int16",
                        name: "dsnpType",
                        type: "int16",
                    },
                    {
                        internalType: "string",
                        name: "url",
                        type: "string",
                    },
                    {
                        internalType: "bytes32",
                        name: "hash",
                        type: "bytes32",
                    },
                ],
                internalType: "struct IPublish.Batch[]",
                name: "publications",
                type: "tuple[]",
            },
        ],
        name: "publish",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IPublish__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
IPublish__factory.abi = _abi;
