// @dsnp/sdk/dist/esm/core/messages/messages.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ethers } from "ethers";
import { requireGetSigner } from "../../config";
import { sortObject } from "../utilities/json";
export var DSNPType;
(function (DSNPType) {
    DSNPType[DSNPType["GraphChange"] = 1] = "GraphChange";
    DSNPType[DSNPType["Broadcast"] = 2] = "Broadcast";
    DSNPType[DSNPType["Reply"] = 3] = "Reply";
    DSNPType[DSNPType["Reaction"] = 4] = "Reaction";
    DSNPType[DSNPType["Profile"] = 5] = "Profile";
})(DSNPType || (DSNPType = {}));
export const createBroadcastMessage = (fromId, url, hash) => ({
    dsnpType: DSNPType.Broadcast,
    contentHash: hash,
    fromId,
    url,
});
export const createReplyMessage = (fromId, url, hash, inReplyTo) => ({
    dsnpType: DSNPType.Reply,
    contentHash: hash,
    fromId,
    inReplyTo,
    url,
});
export const createReactionMessage = (fromId, emoji, inReplyTo) => ({
    dsnpType: DSNPType.Reaction,
    emoji,
    fromId,
    inReplyTo,
});
export var DSNPGraphChangeType;
(function (DSNPGraphChangeType) {
    DSNPGraphChangeType[DSNPGraphChangeType["Follow"] = 1] = "Follow";
    DSNPGraphChangeType[DSNPGraphChangeType["Unfollow"] = 2] = "Unfollow";
})(DSNPGraphChangeType || (DSNPGraphChangeType = {}));
export const createFollowGraphChangeMessage = (fromId, followeeId) => ({
    fromId,
    dsnpType: DSNPType.GraphChange,
    changeType: DSNPGraphChangeType.Follow,
    objectId: followeeId,
});
export const createUnfollowGraphChangeMessage = (fromId, followeeId) => ({
    fromId,
    dsnpType: DSNPType.GraphChange,
    changeType: DSNPGraphChangeType.Unfollow,
    objectId: followeeId,
});
export const serialize = (message) => {
    const sortedMessage = sortObject(message);
    let serialization = "";
    for (const key in sortedMessage) {
        serialization = `${serialization}${key}${sortedMessage[key]}`;
    }
    return serialization;
};
export const createProfileMessage = (fromId, url, hash) => ({
    dsnpType: DSNPType.Profile,
    contentHash: hash,
    fromId,
    url,
});
export const sign = (message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = requireGetSigner(opts);
    const signature = yield signer.signMessage(serialize(message));
    return Object.assign(Object.assign({}, message), { signature });
});
export const recoverPublicKey = (message, signature) => {
    return ethers.utils.verifyMessage(serialize(message), signature);
};
