// @dsnp/sdk/dist/cjs/core/messages/messages.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.recoverPublicKey = exports.sign = exports.createProfileMessage = exports.serialize = exports.createUnfollowGraphChangeMessage = exports.createFollowGraphChangeMessage = exports.DSNPGraphChangeType = exports.createReactionMessage = exports.createReplyMessage = exports.createBroadcastMessage = exports.DSNPType = void 0;
const ethers_1 = require("ethers");
const config_1 = require("../../config");
const json_1 = require("../utilities/json");
var DSNPType;
(function (DSNPType) {
    DSNPType[DSNPType["GraphChange"] = 1] = "GraphChange";
    DSNPType[DSNPType["Broadcast"] = 2] = "Broadcast";
    DSNPType[DSNPType["Reply"] = 3] = "Reply";
    DSNPType[DSNPType["Reaction"] = 4] = "Reaction";
    DSNPType[DSNPType["Profile"] = 5] = "Profile";
})(DSNPType = exports.DSNPType || (exports.DSNPType = {}));
const createBroadcastMessage = (fromId, url, hash) => ({
    dsnpType: DSNPType.Broadcast,
    contentHash: hash,
    fromId,
    url,
});
exports.createBroadcastMessage = createBroadcastMessage;
const createReplyMessage = (fromId, url, hash, inReplyTo) => ({
    dsnpType: DSNPType.Reply,
    contentHash: hash,
    fromId,
    inReplyTo,
    url,
});
exports.createReplyMessage = createReplyMessage;
const createReactionMessage = (fromId, emoji, inReplyTo) => ({
    dsnpType: DSNPType.Reaction,
    emoji,
    fromId,
    inReplyTo,
});
exports.createReactionMessage = createReactionMessage;
var DSNPGraphChangeType;
(function (DSNPGraphChangeType) {
    DSNPGraphChangeType[DSNPGraphChangeType["Follow"] = 1] = "Follow";
    DSNPGraphChangeType[DSNPGraphChangeType["Unfollow"] = 2] = "Unfollow";
})(DSNPGraphChangeType = exports.DSNPGraphChangeType || (exports.DSNPGraphChangeType = {}));
const createFollowGraphChangeMessage = (fromId, followeeId) => ({
    fromId,
    dsnpType: DSNPType.GraphChange,
    changeType: DSNPGraphChangeType.Follow,
    objectId: followeeId,
});
exports.createFollowGraphChangeMessage = createFollowGraphChangeMessage;
const createUnfollowGraphChangeMessage = (fromId, followeeId) => ({
    fromId,
    dsnpType: DSNPType.GraphChange,
    changeType: DSNPGraphChangeType.Unfollow,
    objectId: followeeId,
});
exports.createUnfollowGraphChangeMessage = createUnfollowGraphChangeMessage;
const serialize = (message) => {
    const sortedMessage = json_1.sortObject(message);
    let serialization = "";
    for (const key in sortedMessage) {
        serialization = `${serialization}${key}${sortedMessage[key]}`;
    }
    return serialization;
};
exports.serialize = serialize;
const createProfileMessage = (fromId, url, hash) => ({
    dsnpType: DSNPType.Profile,
    contentHash: hash,
    fromId,
    url,
});
exports.createProfileMessage = createProfileMessage;
const sign = (message, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const signer = config_1.requireGetSigner(opts);
    const signature = yield signer.signMessage(exports.serialize(message));
    return Object.assign(Object.assign({}, message), { signature });
});
exports.sign = sign;
const recoverPublicKey = (message, signature) => {
    return ethers_1.ethers.utils.verifyMessage(exports.serialize(message), signature);
};
exports.recoverPublicKey = recoverPublicKey;
