// @dsnp/test-generators/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sample = exports.randInt = exports.generateBase64String = exports.generateHash = exports.generateUint8Array = exports.generateHexString = exports.addresses = exports.sampleText = void 0;
const sampleTextImport = __importStar(require("./sampleText"));
exports.sampleText = sampleTextImport;
const addressesImport = __importStar(require("./addresses"));
exports.addresses = addressesImport;
const generateHexString = (length) => {
    return "0x" + [...Array(length)].map(() => exports.randInt(16).toString(16)).join("");
};
exports.generateHexString = generateHexString;
const generateUint8Array = (length) => {
    const res = new Uint8Array(length);
    return res.map(() => exports.randInt(256));
};
exports.generateUint8Array = generateUint8Array;
const generateHash = () => {
    return exports.generateHexString(64);
};
exports.generateHash = generateHash;
const generateBase64String = (length) => {
    const randomChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let result = "";
    for (let i = 0; i < length; i++) {
        result += randomChars.charAt(exports.randInt(randomChars.length));
    }
    result += "=";
    return result;
};
exports.generateBase64String = generateBase64String;
const randInt = (max) => {
    return Math.floor(Math.random() * Math.floor(max));
};
exports.randInt = randInt;
const sample = (fromList) => {
    if (fromList.length === 0) {
        throw new Error("cannot sample from an empty list");
    }
    return fromList[Math.floor(Math.random() * fromList.length)];
};
exports.sample = sample;
