// @dsnp/test-generators/dist/cjs/sampleText.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prefabVideos = exports.prefabURLs = exports.prefabMessages = exports.randImage = exports.prefabLastNames = exports.prefabFirstNames = void 0;
exports.prefabFirstNames = [
    "Arna",
    "Tom",
    "安",
    "계향",
    "Ang",
    "Ahsha",
    "Alva",
    "Bob",
    "Bai",
    "Eskil",
    "Bhuvana",
    "Fatima",
    "Duncan",
    "Frank",
    "Elvis",
    "Goins",
    "Josh",
    "Лев Николаевич",
    "Mireille",
    "Cathy",
    "Hamza",
    "Davonte",
    "Petru",
    "Inigo",
    "Norbert",
    "Mary",
    "Beth",
    "Devorah",
    "Marisol",
    "Tala",
    "Katie",
    "Sarah",
    "Srikanth",
    "Elizabeth",
    "Maria",
    "Vállju",
];
exports.prefabLastNames = [
    "Smith",
    "Peterson",
    "Dalisay",
    "Sullivan",
    "Myers",
    "Franklin, Jr.",
    "Fudduni",
    "Grímsdóttir",
    "Khaldun",
    "Cooper",
    "Poitier",
    "Williams",
    "Kotter",
    "Kaur",
    "Jung",
    "Morgan",
    "Begay",
    "Li",
    "장",
    "Montoya",
    "Greenberg",
    "Hämäläinen",
    "Foster",
    "Ng",
    "Srinivasan",
    "Толстой",
    "Boudreau",
    "Sanchez",
    "正文",
];
exports.randImage = "https://picsum.photos/200/300";
exports.prefabMessages = [
    "Hola, mundo!",
    "My favorite cartoon is Spongebob",
    "2020 was the friggin worst.",
    "您所有的基地都属于我们",
    "Go sports team! Dunk the goal!",
    "I am going to the mall tomorrow",
    "Leave me alone I am the grumpiest person",
    "This is the way",
    "Who's on first, what's on Second, I don't know is on third",
    "Beam me up Scotty.",
    "该死的吉姆，我是医生，不是机械师！",
    "Черт возьми, Джим, я врач, а не психиатр!",
    "The pizza shop down the street is giving out free donuts. Kinda sketchy",
    "My birthday is coming up in a little under a month!",
    "I want everyone to know that dogs > cats",
    "Cereal is salad. End of story.",
    "Сухие завтраки - это разновидность салата. Конец истории.",
    "I am highly in favor of ${CITYNAME} ${SPORTSTEAM}. I fervently wish for their continued success in ${SPORTSBALL}",
    "Hot take: Hotdogs aren't a sandwich, they're a taco",
    "It is time for you to look inward, and start asking " +
        "yourself the big questions. Who are you? And what do " +
        "you want?",
    "“Let your plans be dark and impenetrable as night, and when you move, fall like a thunderbolt.”\n" +
        "―Sun Tzu, The Art of War",
    "My favorite sea shanty is 'Friggin in the Riggin'",
    "Darn it, Jim, I'm a family doctor, not a swearing doctor! For pity's sake Jim, this is prime time!",
    "Sometimes life is like this dark tunnel. You can’t always " +
        "see the light at the end of the tunnel, but if you just " +
        "keep moving, you will come to a better place.",
    "Did you ever hear the tragedy of Darth Plagueis the Wise? " +
        "I thought not. It's not a story the Jedi would tell you. " +
        "It's a Sith legend. Darth Plagueis was a Dark Lord of the " +
        "Sith, so powerful and so wise he could use the Force to " +
        "influence the midichlorians to create life... He had such " +
        "a knowledge of the dark side that he could even keep the " +
        "ones he cared about from dying. The dark side of the Force " +
        "is a pathway to many abilities some consider to be unnatural. " +
        "He became so powerful... the only thing he was afraid of was " +
        "losing his power, which eventually, of course, he did. " +
        "Unfortunately, he taught his apprentice everything he knew, " +
        "then his apprentice killed him in his sleep. It's ironic he " +
        "could save others from death, but not himself.",
];
exports.prefabURLs = [
    "https://mydreamsymbolism.com/wp-content/uploads/2017/11/Dreams-About-Babies-Meaning-and-Interpretation.jpg",
    "https://upload.wikimedia.org/wikipedia/commons/2/20/Baby_2.jpg",
    "https://live.staticflickr.com/6064/6031231222_eea9093801_b.jpg",
    "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT1B8ZsbUPkarzs2ZsMTpAoICuqPiFrEIrKyQ&usqp=CAU",
    "https://image.shutterstock.com/image-photo/happy-african-american-little-girl-260nw-753499141.jpg",
    "https://live.staticflickr.com/8210/8165506503_31105e0ece_b.jpg",
    "https://upload.wikimedia.org/wikipedia/commons/6/64/The_Puppy.jpg",
    "https://s2.best-wallpaper.net/wallpaper/iphone/1407/Little-dog-black-puppy_iphone_640x1136.jpg",
    "https://p2.piqsels.com/preview/947/498/380/puppy-running-dog-animal.jpg",
    "https://be.chewy.com/wp-content/uploads/sleeping-kitten-TS_160626325-1.jpg",
    "https://thumbs.dreamstime.com/b/black-little-kitten-walking-23747329.jpg",
    "https://www.advantagepetcare.com.au/sites/g/files/adhwdz311/files/styles/paragraph_image/public/2020-07/istock-871246578_unrestricted_1110x800.jpg?itok=wuuBaqf-",
    "https://image.shutterstock.com/image-photo/british-blue-kitten-very-beautiful-260nw-796071583.jpg",
    "https://media.istockphoto.com/photos/bengal-kitten-in-flower-meadow-picture-id905117504?b=1&k=6&m=905117504&s=612x612&w=0&h=wN_f34X4QvADggBid_58qUCgrOU8oE9SuLWwgP0IKgE=",
    "https://www.theactuary.com/sites/default/files/images/121912_1.jpg",
];
exports.prefabVideos = [
    "https://youtu.be/ZYBliPhLAE0",
    "https://vimeo.com/25436623",
    "https://www.youtube.com/watch?v=6r7TqPNxhrM",
    "https://vimeo.com/479212171",
    "https://www.youtube.com/watch?v=k_IAH7D-vTI",
    "https://www.youtube.com/watch?v=AVoHlghb0CM",
    "https://www.facebook.com/geetgandhar/videos/941987132933178/",
    "https://www.youtube.com/watch?v=ywWBy6J5gz8",
    "https://www.youtube.com/watch?v=Ns4TPTC8whw",
    "https://www.youtube.com/watch?v=lyZQPjUT5B4",
    "https://www.youtube.com/watch?v=ROalU379l3U",
    "https://www.youtube.com/watch?v=XaqR3G_NVoo",
    "https://www.youtube.com/watch?v=iP897Z5Nerk",
    "https://www.youtube.com/watch?v=CmPA7zE8mx0",
];
