// @dsnp/test-generators/dist/esm/index.js
import * as sampleTextImport from "./sampleText";
export const sampleText = sampleTextImport;
import * as addressesImport from "./addresses";
export const addresses = addressesImport;
export const generateHexString = (length) => {
    return "0x" + [...Array(length)].map(() => randInt(16).toString(16)).join("");
};
export const generateUint8Array = (length) => {
    const res = new Uint8Array(length);
    return res.map(() => randInt(256));
};
export const generateHash = () => {
    return generateHexString(64);
};
export const generateBase64String = (length) => {
    const randomChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let result = "";
    for (let i = 0; i < length; i++) {
        result += randomChars.charAt(randInt(randomChars.length));
    }
    result += "=";
    return result;
};
export const randInt = (max) => {
    return Math.floor(Math.random() * Math.floor(max));
};
export const sample = (fromList) => {
    if (fromList.length === 0) {
        throw new Error("cannot sample from an empty list");
    }
    return fromList[Math.floor(Math.random() * fromList.length)];
};
